Explorer/Assets/Scripts/CRDT/Protocol/CRDTProtocol.cs

```diff
@@ -0,0 +1,305 @@
+                return ProcessMessageResult.StateOutdatedTimestamp;
+
+            var compareDataResult = CRDTMessageComparer.CompareData(in storedData.Data, message.Data);
+
+            switch (compareDataResult)
+            {
+                case 0:
+                    // Right the same message, nothing to do
+                    return ProcessMessageResult.NoChanges;
+                case > 0:
+                    // Nothing to do
+                    return ProcessMessageResult.StateOutdatedData;
+                default:
+                    UpdateLWWState(true, true, inner, in message, ref storedData);
+
+                    // The local state is updated
+                    return ProcessMessageResult.StateUpdatedData;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void UpdateLWWState(bool innerSetExists, bool componentExists, PooledDictionary<CRDTEntity, EntityComponentData> inner,
+            in CRDTMessage crdtMessage, ref EntityComponentData componentData)
+        {
+            if (!innerSetExists)
+                crdtState.lwwComponents[crdtMessage.ComponentId] = inner = new PooledDictionary<CRDTEntity, EntityComponentData>();
+
+            if (!componentExists)
+                crdtState.messagesCount++;
+
+            UpdateLWWState(crdtMessage.Timestamp, crdtMessage.Data, ref componentData);
+
+            // We don't have a ref-wise API for Dictionary Values so we have to assign the data back
+            inner[crdtMessage.EntityId] = componentData;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void UpdateLWWState(int timestamp, ReadOnlyMemory<byte> data, ref EntityComponentData componentData)
+        {
+            componentData.Timestamp = timestamp;
+            componentData.Data = data;
+        }
+
+        private void DeleteEntity(CRDTEntity entityId, int entityNumber, int entityVersion, bool entityWasDeleted)
+        {
+            crdtState.deletedEntities[entityNumber] = entityVersion;
+
+            if (!entityWasDeleted)
+                crdtState.messagesCount++;
+
+            foreach (var componentsStorage in crdtState.lwwComponents.Values)
+            {
+                if (componentsStorage.Remove(entityId))
+                    crdtState.messagesCount--;
+            }
+
+            foreach (var componentsSet in crdtState.appendComponents)
+            {
+                if (componentsSet.Value.TryGetValue(entityId, out var list))
+                {
+                    crdtState.messagesCount -= list.Count;
+                    list.Dispose();
+                    componentsSet.Value.Remove(entityId);
+                }
+            }
+        }
+
+        /// <summary>
+        /// The execution is expensive, consider invoking from a background thread only
+        /// </summary>
+        private bool TryAppendComponent(in CRDTMessage message)
```

Comment:
We have allocations for creating `PooledList` and `PooledDictionary` but only for wrappers as they use `ArrayPool` for internal buffers. I would like to try this approach because of simplicity, if there will be too many allocations for these wrappers we can use Pools of `List` and `Dictionary` instead, or even a pool for `PooledList` and `PooledDictionary`